arbnode/transaction_streamer.go

```diff
@@ -1218,7 +1218,32 @@ func (s *TransactionStreamer) executeMessages(ctx context.Context, ignored struc
 	return s.config().ExecuteMessageLoopDelay
 }
 
-func (s *TransactionStreamer) Start(ctxIn context.Context) error {
+func (s *TransactionStreamer) Start(ctxIn context.Context, syncTillBlock uint64) error {
 	s.StopWaiter.Start(ctxIn, s)
-	return stopwaiter.CallIterativelyWith[struct{}](&s.StopWaiterSafe, s.executeMessages, s.newMessageNotifier)
```

Comment:
syncTillBlock should come from config and not from start.
Also, instead of a weird wrapper - much better to introduce this logic into sub functions.
In executeMessages - no need to do anything if past syncTillBlock

I suspect that the best place to stop adding new messages is in writeMessages - you could return an error if pos is > syncTillBlock. This could be a specific error, and all those other sequencing location could check for it when their sequencing fails instead of having their separate syncTillBlock logic.
There is an edge case in writeMessages if pos < syncTillBlock but pos+len(messages)>syncTillBlock - I think that in that case you should not return an error. This is a debug functionality and it's o.k. if we got a few extra blocks in, but not great if we're missing more blocks than we should.

Reply:
Done.